agreement/events.go

```diff
@@ -942,7 +988,12 @@ func (e checkpointEvent) AttachConsensusVersion(v ConsensusVersionView) external
 }
 
 func (e messageEvent) AttachValidatedAt(d time.Duration) messageEvent {
-	e.Input.Proposal.validatedAt = d
+	if e.T == payloadVerified {
+		e.Input.Proposal.validatedAt = d
+	} else if e.T == voteVerified && e.Input.UnauthenticatedVote.R.Step == 0 {
+		// if this is a proposal vote (step 0), record the validatedAt time on the vote
```

Comment:
I think that attaching it to votes of any step might be okay.
The overhead of reading the clock is already there, so just storing it along with the vote doesn't seem too much of an overhead. It also keeps it more consistent (I can now read the time any vote was verified, no matter the step)